### DISCUSSION TITLE
Unsure on how to deal with rate limiting

### DISCUSSION THREAD

**Discussion by BlossomiShymae** [TheDrone7/shieldbow#8] [Q&A] [answered]
Hi,

I have been wanting to crawl matches using this library in development but I'm facing this issue with the rate limiting.

I made a minimum reproducible example that makes 102 requests, if my understanding is correct. One for the Summoner, one for the MatchId string list, and a hundred for all the Match data.
```typescript
  const shieldbow = new Client(`${RIOT_API_KEY}`);

  await shieldbow.initialize({
    region: "na",
  });
  const summoner = await shieldbow.summoners.fetchBySummonerName(
    "time for uwu"
  );
  const matchIds = await shieldbow.matches.fetchMatchListByPlayer(summoner, {
    queue: 450,
    count: 100,
  });
  const matches = await matchIds.reduce<Promise<Match[]>>(
    async (previous, id) => {
      const _ = await previous;
      return [..._, await shieldbow.matches.fetch(id)];
    },
    Promise.resolve([])
  );

  console.log(matches.length);
```

I'm also using a development key, so my rate limits are 20 requests per second and 100 requests per 2 minutes. 

Running the example above gives me this error.
```
ApiError: 429:: Too Many Requests
You have reached the rate limit.

Retry in -18768ms.
    at C:\Users\Maevie\Desktop\shieldbow-sandbox\node_modules\shieldbow\dist\api\handler.js:55:24
    at new Promise (<anonymous>)
    at ApiHandler.makeApiRequest (C:\Users\Maevie\Desktop\shieldbow-sandbox\node_modules\shieldbow\dist\api\handler.js:45:16)
    at C:\Users\Maevie\Desktop\shieldbow-sandbox\node_modules\shieldbow\dist\managers\MatchManager.js:51:22
    at new Promise (<anonymous>)
    at MatchManager.fetch (C:\Users\Maevie\Desktop\shieldbow-sandbox\node_modules\shieldbow\dist\managers\MatchManager.js:46:16)
    at C:\Users\Maevie\Desktop\shieldbow-sandbox\index.js:34:47
    at Generator.next (<anonymous>)
    at fulfilled (C:\Users\Maevie\Desktop\shieldbow-sandbox\index.js:5:58)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  request: 'Get Match By Match ID (Match ID: NA1_4346679256)',
  response: {
    data: null,
    status: 429,
    statusText: 'Too Many Requests',
    headers: {},
    config: {}
  }
}
Done in 110.40s.
```

I have read the documentation and know that the rate limiter itself has caveats. I expected the rate limit to be reached (102 requests exceeding 100 within 2 minutes) but not the way it was dealt with. I assumed it would respect the rate limit from the HTTP response headers before getting the next request (by waiting and not failing fast).

How should I approach this problem? I'm thinking of using a backoff and retry but I don't see a way to get the retry time from the Ratelimiter class when this error occurs. Should I just use rate limiting from another library for the mean time? Sorry if I'm not understanding this at all >.<

### Information
- OS: Windows 10
- Node: v16.15.1
- Shieldbow: v1.3.0

**Comment by BlossomiShymae** [comment, marked as answer]
I ended up answering my problem with a short-term solution in the meantime. I used node-rate-limiter and followed this [guide](https://www.useanvil.com/blog/engineering/throttling-and-consuming-apis-with-429-rate-limits/) to implement a wrapper function for each api call made with shieldbow. The limiter is shared by dependency injection across my functions and controllers.

For some reason I could not get Promise.all to work when dealing with large match lists so I ended up using a Promise Array.reduce for serial execution. I now have no problem with crawling matches for testing! :bee_happy:
```Typescript
    // Example excerpt
    const player = await limiter.acquireToken<Summoner>(() =>
      shieldbow.summoners.fetchBySummonerName(name)
    );
    const options: MatchByPlayerOptions = {
      queue: 450,
      count: 100,
      start: start,
    };
    const matchIds = await limiter.acquireToken<string[]>(() =>
      shieldbow.matches.fetchMatchListByPlayer(player, options)
    );

    const matches = await matchIds.reduce<Promise<Match[]>>(
      async (previous, id) => {
        const _ = await previous;
        const match = await limiter.acquireToken<Match>(() =>
          shieldbow.matches.fetch(id)
        );
        return [..._, match];
      },
      Promise.resolve([])
    );
```

EDIT 1/19/2023:
This is a version of a retry wrapper function I made that utilizes the retry duration in the rejection message thrown by `shieldbow` instead of using hardcoded limits. This breaks with `Promise.All()` but if used properly (segregating promise calls by region/endpoint) then it *should* worksies. :3
```Typescript
const requestRetryOrThrow = async <T>(
    action: () => Promise<T>,
    attempts: number = 3
  ) => {
    let durationMilliseconds = 15 * 1000;
    let attempted = 0;
    while (true) {
      try {
        let result = await action();
        return result;
      } catch (e) {
        if (attempted >= attempts) throw new Error("Exceeded max attempts");
        if (e instanceof Error) {
          if (e.message.includes("429")) {
            const parsedMilliseconds = /-(.*?)ms/gm.exec(e.message);
            // Unable to parse. Real 429 or changed message, exit immediately! :c
            if (parsedMilliseconds === null)
              throw new Error("Failed to parse retry backoff from message");
            durationMilliseconds = parseInt(parsedMilliseconds[1], 10);
          } else {
            // Not implemented (4xx, 5xx...)
            throw e;
          }
          console.log(
            `Retrying after ${durationMilliseconds / 1000} seconds...`
          );
          await new Promise((resolve) =>
            setTimeout(resolve, durationMilliseconds)
          );
        }
        attempts++;
      }
    }
  };
```

**Comment by TheDrone7** [comment]
BTW, this lib does have a half-assed rate limiter (I want to add a custom rate-limiter instead of just using node-rate-limiter).

Even though you see the rate limited errors, it's not actually sending those requests to the RIOT API to prevent you from losing your key.

**Comment by SalmonMode** [comment]
If it helps, my approach has two pieces:

1) I modified version of the repo to give the match manager the same storage system as some other managers so it could really on previously stored matches in repeat runs.

2) I used async generators when crawling over matches. Every time it "fetches" a match, it yields it to whatever is calling it (probably in an async for loop) so it can be processed, making use of the time it'll possibly have to wait before it can make another API request.

The async generator function I made takes a gets the match IDs for a summoner and makes sure to track the time it last made an API request. While iterating over the match ID list, it first checks if the match exists in storage (there might be a better way to make sure caches and storage work together nicely), and if it does, it yields that match. If it doesn't exist in storage, though, it does some quick math to find out how long it needs to wait before it can safely make the next request (1200ms between requests), awaits a simple delay function for how long it needs to wait (some amount less than 1200ms), then fetches the match from the API (updating the last request time for future iterations), then yields the match.

I'll update with a code snippet in a little bit to help, as that's some lovely word salad haha. But maybe the client could provide such an async generator function, or at least provide this a passive rate limiting option. Having an async generator example in the docs might also be helpful for folks less familiar with JavaScript to help them get a bit more performance (I'd be happy to write it up).

Another, possibly more comprehensive rate limiting option could be to not rate limit until 100 requests have been made in the past 2 minutes, and then delay the next request until 2 minutes and 1.2 seconds after the first request from those 2 minutes was made.

A queue of timestamps with a max size of 100 could work nicely for this, I think. Every time a request is made, the client could dequeue any timestamps older than 2 minutes and 1.2 seconds, not allow more than 100 items in the queue, and delay any requests until it can create space in the queue.

**Reply by SalmonMode** [reply]
Example:

```typescript
/**
 * Given a time, will return a Promise that takes that long to resolve.
 *
 * Can be used to delay execution by a specific amount of time.
 *
 * @param timeInMilliseconds time to wait in milliseconds
 */
async function delay(timeInMilliseconds: number): Promise<void> {
  await new Promise(resolve => setTimeout(resolve, timeInMilliseconds));
}

async function getTotalMatchPlayTimeForSummoner(summoner: Summoner, client: Client): Promise<number> {
  let totalPlayTimeInMilliseconds: number = 0;

  let matchHistoryGenerator = getMatchHistoryForSummoner(summoner, client);
  for await (let match of matchHistoryGenerator) {
    // Could be doing potentially time consuming operation that burns time that might otherwise
    // be spent doing nothing while waiting to make the next request
    totalPlayTimeInMilliseconds += match.endTimestamp - match.startTimestamp;
  }
  return totalPlayTimeInMilliseconds;
}

async function* getMatchHistoryForSummoner(summoner: Summoner, client: Client): AsyncIterable<Match> {
  let matchIds = await getMatchIdHistoryForSummoner(summoner);
  // Track the last time a request was made. This does not fully account for any API
  // requests made by getMatchIdHistoryForSummoner.
  let lastRequestTime: number = Date.now();
  for (let matchId of matchIds) {
    try {
      // Check if match was already stored by storage manager. Must be checked before
      // calling client.matches.fetch so we can know whether or not an API request will
      // be made.
      let matchWasAlreadyStored = await client.matches.hasMatchStored(matchId);
      if (!matchWasAlreadyStored) {
        // The client is about to make an API request, so should delay to avoid rate limiter.
        let timeDiff = Date.now() - lastRequestTime;
        await delay(1200 - timeDiff);
      }
      // client.matches.fetch is already modified to use storage manager and will use stored
      // matches if available, rather than making requests to the API
      let matchDetails = await client.matches.fetch(matchId);
      if (!matchWasAlreadyStored) {
        // made a request, so update the last request time so future iterations can avoid
        // hitting the rate limiter.
        lastRequestTime = Date.now();
      }
      // Yield the match details so they can be processed.
      yield matchDetails;
    } catch (error) {
      // If the match couldn't be processed, just skip it.
      console.log(`Match (${matchId}) could not be fetched`);
      console.log(error);
      console.log(error.message);
      continue;
    }
  }
}
async function getMatchIdHistoryForSummoner(summoner: Summoner): Promise<string[]> {
  let matchIdStrings: string[] = [];
  // try going back 10 pages of match history to get everything.
  for (let i = 0; i < 10; i++) {
    let ids: string[] = await summoner.fetchMatchList({ start: i * 100, count: 100 });
    matchIdStrings.push(...ids);
  }
  return matchIdStrings;
}
```
